external/java/src/main/java/ch/epfl/dedis/lib/crypto/bn256/BN.java

```diff
@@ -0,0 +1,455 @@
+            this.k = k;
+            this.p = p;
+        }
+    }
+
+    /**
+     * The pair of a G2 point and a scalar that is used to generate it.
+     */
+    public static class PairG2 {
+        public BigInteger k;
+        public G2 p;
+        public PairG2(BigInteger k, G2 p) {
+            this.k = k;
+            this.p = p;
+        }
+    }
+
+    /**
+     * A point in G1.
+     */
+    public static class G1 {
+        CurvePoint p;
+
+        /**
+         * Construct a G1 point. There is no guarantee on its value, please set it later.
+         */
+        public G1() {
+            this.p = new CurvePoint();
+        }
+
+        /**
+         * Construct a G1 point using a given curve point.
+         * @param p is the curve point.
+         */
+        public G1(CurvePoint p) {
+            this.p = p;
+        }
+
+        /**
+         * Generate a random pair of a point and a scalar that is used to create the point.
+         * @param rnd is the random source.
+         * @return a pair of a point and a scalar.
+         */
+        public static PairG1 rand(Random rnd) {
+            BigInteger k = randPosBigInt(rnd, Constants.order);
+            G1 p = new G1().scalarBaseMul(k);
+            return new PairG1(k, p);
+        }
+
+        public String toString() {
+            return "bn256.G1" + this.p.toString();
+        }
+
+        /**
+         * Perform a scalar multiplication with the generator point.
+         * @param k is the scalar.
+         * @return the result which is also this object.
+         */
+        public G1 scalarBaseMul(BigInteger k) {
+            this.p.mul(CurvePoint.curveGen, k);
+            return this;
+        }
+
+        /**
+         * Perform a scalar multiplication.
+         * @param a is the point.
+         * @param k is the scalar.
+         * @return the result which is also this object.
+         */
+        public G1 scalarMul(G1 a, BigInteger k) {
+            this.p.mul(a.p, k);
+            return this;
+        }
+
+        /**
+         * Perform a point addition.
+         * @param a is a point.
+         * @param b is a point.
+         * @return the resulting point which is also this object.
+         */
+        public G1 add(G1 a, G1 b) {
+            this.p.add(a.p, b.p);
+            return this;
+        }
+
+        /**
+         * Perform a point negation.
+         * @param a is the point for negation.
+         * @return the resulting point which is also this object.
+         */
+        public G1 neg(G1 a) {
+            this.p.negative(a.p);
+            return this;
+        }
+
+        /**
+         * Turns the point into its byte representation.
+         * @return the marshalled bytes.
+         */
+        public byte[] marshal() {
+            final int numBytes = 256/8;
+
+            if (this.p.isInfinity()) {
+                return new byte[numBytes*2];
+            }
+
+            this.p.makeAffine();
```

Comment:
I prefer calling the `makeAffine` on a copy, else you 'change' the underlying point. Even though it should only be changed in a way that the point is 'identical', it still changes it.

Or perhaps that's just my experience with go-races in this code...

Reply:
Good point, this actually touches on a bigger problem. If we want to make it fully thread-safe, we need to make sure no operation can change the underlying point (at the moment every operation changes the inner points/elements). Effectively, making the classes immutable (final classes), just like `BigInteger`.

I feel there are pros/cons to this approach. Making it thread-safe will make the package look and feel quite different from the one in golang/crypto and it'll be harder to port future changes. Keeping the classes mutable as the one in golang/crypto will be easier to maintain. We can write a comment and say all the classes in this package are not thread-safe and users should take appropriate measures (e.g., locking/copying).

On the other hand, if our clients are very multi-threaded, then we definitely need to make the classes final. What are your opinions @jeffallen @Gilthoniel ?

Reply:
I'm facing race conditions because of this approach in the go code so I would go in Linus' direction. I'm not sure we need to have the whole class immutable, in my opinion that's better but if you say that it will be hard to maintain, it's probably not worth. The real issue is this makeAffine that is used in stuff you're not expecting writes from it, like the marshal function. I'd change only this to return the affine result instead of mutate the class if that makes sense.

Reply:
1. Probably no big changes
2. I can't use `dedis/kyber` as reference because it's using assembly, so the internal implementation is very different and you cannot use it as a reference for porting to java. `golang/crypto` only uses `big.Int` so it's much easier to port.
3. I'm sure it will be superseded, but I doubt that will be soon. Right now we don't have a candidate curve as far as I know. When there is one, we cannot use it immediately because it might not be secure. We'll have to wait for the security research community to scrutinize it.

Reply:
> I'm facing race conditions because of this approach in the go code so I would go in Linus' direction. I'm not sure we need to have the whole class immutable, in my opinion that's better but if you say that it will be hard to maintain, it's probably not worth. The real issue is this makeAffine that is used in stuff you're not expecting writes from it, like the marshal function. I'd change only this to return the affine result instead of mutate the class if that makes sense.

I will look at whether it's easy to make the class final.